### ISSUE TITLE
Bug (Keyboard Accessibility): Add "Skip to main content" link

### ISSUE THREAD

**Issue by dsmchen** [TheOdinProject/theodinproject#5199]
### Checks

- [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [x] The title of this issue follows the `Bug: brief description of bug` format, e.g. `Bug: Lesson complete button does not update on click`
- [x] Would you like to work on this issue?

### Bug description

- Parent issue: https://github.com/TheOdinProject/theodinproject/issues/2893
- "On most pages, keyboard and screen reader users must navigate a long list of navigation links and other elements before ever arriving at the main content ... Skip navigation links are useful to give screen reader and keyboard users the same capability of navigating directly to the main content." (https://webaim.org/techniques/skipnav/#overview)

### How to reproduce

1. Go to https://www.theodinproject.com/
2. Press tab key
3. First tab keypress goes to header logo anchor element

<img width="625" height="116" alt="Image" src="https://github.com/user-attachments/assets/92fd1471-77eb-4372-a656-70804a7416bf" />

### Expected behavior

1. Go to https://www.theodinproject.com/
2. Press tab key
3. First tab keypress goes to "Skip to main content" link

<img width="514" height="137" alt="Image" src="https://github.com/user-attachments/assets/73faa91f-c1fd-40f5-a201-0e978e00fd07" />

### What browsers are you seeing the problem on?

Firefox, Chrome

### What OS are you using?

MacOS

### Discord Name

doris6601

**Comment by dsmchen** [comment]
Hey @mao-sz, what does this issue need to remove "Status: Needs Review"?

**Comment by mao-sz** [comment]
It first needs me to actually send the message I thought I had sent but clearly didn't 🤦

@thatblindgeye any thoughts on where might be the most appropriate place in the DOM for this kind of thing, if we did include it? Not sure how that might play with the various different ToC options being discussed. Is this something that requires figuring out the ToC stuff first before we have an idea for the "skip to main content" button?

**Comment by dsmchen** [comment]
Hi @thatblindgeye, in case you missed the above message :)

**Comment by dsmchen** [comment]
Hi @thatblindgeye, following up on the above :)

**Comment by mao-sz** [comment]
@dsmchen Had some time to revisit this a little more deeply. Is this something you're still interested in working on?

An invisible-until-focused skip link makes sense but on top of actually making the skip link, we'd need to add an appropriate ID like `id="main-content"` to the main content element for the link to href to. That involves a few things:
- Some pages will have the main content inside a `<div>` instead of a `<main>`. Such pages would ideally have this `<div>` changed to a `<main>`, and the right ID given to it.
- Some pages have multiple `<section>`s but they're not all contained within a `<main>` or equivalent (i.e. they're siblings to the `<nav>`).
- Speaking of the `<nav>`, it's not within a `<header>` and I think we ought to have the appropriate landmarks (header -> main -> footer). We have the footer, but the header and main aren't actually in those landmarks and should be.

Which means to implement the skip link, we'd need to:
1. Put the `<nav>` in a `<header>`.
2. Make sure all pages' main content is inside a `<main>`. That will involve amending views for lesson pages but also any of [our static pages](https://github.com/TheOdinProject/theodinproject/tree/main/app/views/static_pages), and may involve wrapping things in a `<main id="main-content">` or changing a "main div" to a `<main>` with the ID.
3. *Then* we should be able to add the invisible-until-focused skip link to the header with `href="#main-content"` as the first focusable.
4. We'd also need to ensure that any relevant selectors in CSS/client-side JS are amended due to the new layout.
5. Write tests.

@KevinMulhern does this sound about right to you? Anything I may have missed, particularly with specifics of this code base?

If we do come to think about redesigning the lesson ToC then we can always work around what we have at the time, rather than having that discussion block this issue.

**Comment by dsmchen** [comment]
@mao-sz Thank you for the detailed comment! 👍 I'm still interested in working on this. However, I'd like to complete the React course first, I'm currently on https://www.theodinproject.com/lessons/node-path-react-new-fetching-data-in-react#assignment. If time is a priority, then I'm happy for someone else to work on this.

**Comment by mao-sz** [comment]
No worries, this isn't urgent. Still would like a +1 on things that'd need to be done by someone more familiar with this codebase.

**Comment by dsmchen** [comment]
> Still would like a +1 on things that'd need to be done by someone more familiar with this codebase.

@mao-sz Is this a blocker?

**Comment by mao-sz** [comment]
Yes it is. I'm not familiar enough with this codebase to sign off on that acceptance criteria. It's also not urgent so I'm in no rush to get this shipped.

**Comment by dsmchen** [comment]
@mao-sz Got it. Thanks for clarifying!

**Comment by KevinMulhern** [comment]
That plan looks solid to me @mao-sz 💪 

For adding main to all pages. The cleanest approach could be adding `<main id="main-content">` in the [application layout, wrapping the `<%= yield %>`](https://github.com/TheOdinProject/theodinproject/blob/main/app/views/layouts/application.html.erb#L38). That way every page gets it automatically without us needing to touch every individual view and save a good bit of work.

**Comment by mao-sz** [comment]
Awesome, so the acceptance criteria would be:

- Wrap `<%= yield %>` in the application layout with a `<main id="main-content">`
- Wrap the header `<nav>` in an actual `<header>`
- Add invisible-until-focused anchor in the header with `href="#main-content"`
- Amend any CSS as necessary
- Write tests as necessary

Let us know if you run into any problems @dsmchen

**Comment by mao-sz** [comment]
@KevinMulhern thoughts on how best to proceed given what we discussed in https://github.com/TheOdinProject/theodinproject/pull/5311#issuecomment-4383996255? Since the sidebar now changes the hierarchy, the AC above is no longer sufficient.

**Comment by dsmchen** [comment]
Hi @mao-sz & @KevinMulhern, sorry I've been MIA, I'm happy to proceed once the AC is confirmed 🫡

**Comment by dsmchen** [comment]
Hi @mao-sz & @KevinMulhern, following up on the above 🙏
